lib/source_file.cpp

```diff
@@ -226,6 +226,36 @@ cstring InputSources::getSourceFragment(const SourceInfo &position) const {
     return result + toadd + marker + cstring::newline;
 }
 
+cstring InputSources::getSourceFragment2(const SourceInfo &position) const {
+    if (!position.isValid())
+        return "";
+
+    cstring result = this->getLine(position.getStart().getLineNumber());
+    unsigned int result_len = strlen(result);
+    unsigned int start = position.getStart().getColumnNumber();
+    unsigned int end, len;
+    cstring toadd = "";
+
+    // If the position spans multiple lines, truncate to just the first line
+    if (position.getEnd().getLineNumber() > position.getStart().getLineNumber()) {
+        // go to the end of the first line
+        end = strlen(result);
+        if (result.find('\n') != nullptr) {
+            --end;
+        }
+        toadd = " ...";
+    } else {
+        end = position.getEnd().getColumnNumber();
+    }
+    len = end - start;
+
+    char substr[result_len+1];
+    memcpy(substr, &result[start], len);
```

Comment:
It'd be better to use [this](https://github.com/p4lang/p4c/blob/master/lib/cstring.h#L147) instead of memcpy.